File: backends/node-ts/src/cursor.ts
import { BadRequestProblem } from "./problems.js";

/**
 * Keyset position for the v2 listing: the `(createdAt, id)` of the last item on
 * the previous page, wrapped in base64url so clients treat it as opaque. Keyset
 * pagination is what makes v2 stable under concurrent inserts — new rows land
 * before the cursor position and cannot shift what the next page returns.
 */
export interface BookmarkCursor {
  createdAt: Date;
  id: string;
}

const UUID_PATTERN = /^[0-9a-f]{8}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{12}$/;

export function encodeCursor(cursor: BookmarkCursor): string {
  return Buffer.from(`${cursor.createdAt.toISOString()}|${cursor.id}`, "utf8").toString("base64url");
}

export function decodeCursor(value: string): BookmarkCursor {
  const decoded = Buffer.from(value, "base64url").toString("utf8");
  const separator = decoded.indexOf("|");
  const createdAt = new Date(decoded.slice(0, separator));
  const id = decoded.slice(separator + 1);
  if (separator < 0 || Number.isNaN(createdAt.getTime()) || !UUID_PATTERN.test(id)) {
    throw new BadRequestProblem("The cursor is malformed or unresolvable.");
  }
  return { createdAt, id };
}

File: backends/go/internal/bookmarks/cursor.go
package bookmarks

import (
	"encoding/base64"
	"strings"
	"time"

	"github.com/google/uuid"

	"github.com/kamkie/stackverse/backends/go/internal/web"
)

// Cursor is the keyset position for the v2 listing: the (createdAt, id) of the
// last item on the previous page, wrapped in base64url so clients treat it as
// opaque. Keyset pagination is what makes v2 stable under concurrent inserts —
// new rows land before the cursor position and cannot shift the next page.
//
// "Unresolvable cursor → 400" is interpreted as *undecodable*: a well-formed
// but never-issued cursor is indistinguishable, on a stateless service, from a
// legitimate cursor whose boundary row was deleted between pages — and that
// one must keep working.
type Cursor struct {
	CreatedAt time.Time
	ID        uuid.UUID
}

func (c Cursor) Encode() string {
	raw := c.CreatedAt.UTC().Format(time.RFC3339Nano) + "|" + c.ID.String()
	return base64.RawURLEncoding.EncodeToString([]byte(raw))
}

func DecodeCursor(cursor string) (Cursor, *web.Problem) {
	malformed := web.BadRequest("The cursor is malformed or unresolvable.")
	decoded, err := base64.RawURLEncoding.DecodeString(cursor)
	if err != nil {
		return Cursor{}, malformed
	}
	createdAt, id, ok := strings.Cut(string(decoded), "|")
	if !ok {
		return Cursor{}, malformed
	}
	parsedTime, err := time.Parse(time.RFC3339Nano, createdAt)
	if err != nil {
		return Cursor{}, malformed
	}
	parsedID, err := uuid.Parse(id)
	if err != nil {
		return Cursor{}, malformed
	}
	return Cursor{CreatedAt: parsedTime, ID: parsedID}, nil
}
